File: 103_Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/*
Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, 
from left to right, then right to left for the next level and alternate between).

For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its zigzag level order traversal as:
[
  [3],
  [20,9],
  [15,7]
]
*/
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
    public:
        struct TreeNode {
            int val;
            TreeNode *left;
            TreeNode *right;
            TreeNode(int x) : val(x), left(NULL), right(NULL) {}
        };

        vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
            if(!root) return vector<vector<int>>();
            vector<vector<int>> res;
            queue<TreeNode*> Q;
            Q.push(root);
            bool L2R = true;
            while(!Q.empty()) {
                int size = Q.size();
                vector<int> row(size);
                for(int i = 0; i < size; i++) {
                    TreeNode* temp = Q.front();
                    Q.pop();
                    int index = L2R ? i : size - 1 - i;
                    row[index] = temp->val;
                    if(temp->left) Q.push(temp->left);
                    if(temp->right) Q.push(temp->right);
                }
                L2R = !L2R;
                res.push_back(row);
            }
            return res;
        }
        
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/103_Binary_Tree_Zigzag_Level_Order_Traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        res, queue = [], [(root, 0)]
        while queue:
            cur, level = queue.pop()
            if cur:
                if len(res) < level + 1:
                    res.append([])
                if level % 2:
                    res[level].insert(0, cur.val)
                else:
                    res[level].append(cur.val)
                queue.append((cur.left, level + 1))
                queue.append((cur.right, level + 1))
        return res
